python/ray/experimental/sgd_api.py

```diff
@@ -0,0 +1,315 @@
+        num_cpus_per_worker: Number of CPUs per worker
+        num_gpus_per_worker: Number of GPUs per worker
+        backend: torch, tensorflow, horovod, distml
+        config: If using training function, config will get passed directly to
+        the training function. If using TrainingOperator class, then it
+        doesn't do anything.
+        """
+        if backend == "horovod":
+            self.actor_group = HorovodActorGroup()
+        ...
+        self.train_fn_or_cls = train_fn_or_cls
+        self.actor_group.start()
+        if not callable(self.train_fn_or_cls):
+            self.actor_group.start(self.train_fn_or_cls)
+        else:
+            # Stateless if using function API
+            self.actor_group.start(None)
+
+    def init_operator(self):
+        """Initialization for class API usage."""
+        pass
+
+    # For function API it is a single run call
+    def run(self):
+        if callable(self.train_fn_or_cls):
+            self.actor_group.run(lambda w: train_fn(self.config))
+
+    # For class API, it's similar to our current TorchTrainer
+
+    def train_epoch(self):
+        if not callable(self.train_fn_or_cls):
+            return self.actor_group.run(lambda w: w.train_epoch())
+
+    def validate(self):
+        if not callable(self.train_fn_or_cls):
+            return self.actor_group.run(lambda w: w.validate())
+
+    def state_dict(self):
+        if not callable(self.train_fn_or_cls):
+            return self.actor_group.run(lambda w: w.state_dict())
+
+    def save_checkpoint(self):
+        if not callable(self.train_fn_or_cls):
+            return self.actor_group.run(lambda w: w.save_checkpoint())
+
+# Function API usage
+def train_func(config, world_rank, local_rank, device_id ):
+    # Either
+    state = ray.sgd.load_checkpoint()
+    model, optimizer, scheduler, train_loader, val_loader = state.model, \
+        state.optimizer, state.scheduler, state.train_loader, state.val_loader
+    # Or
+    model = Net()
+    optimizer = torch.sgd()
+    ...
+
+    # These should work.
+    ray.sgd.local_rank()
+    ray.sgd.world_rank()
+
+    # User sets up mixed precision here.
+
+    model.cuda()
+    DistributedDataParallel(...)
+
+    # Potentially add DistributedSampler here
+    train_loader.sampler = DistributedSampler(...)
+
+
+    # eventually, optional: ray.sgd.accelerate(model, …)
+    for _ in config["num_epochs"]:
+        train(...)
+        # optional, eventually: ray.sgd.report(...)
+        # at some frequency, ray.sgd.save_checkpoint(state, files=False|True)
+        validate(...)
+    return model
+
+trainer = Trainer(train_fn, num_workers=“auto”, backend=...)
+result = Trainer.run(train_func, callbacks=[MLflowLogger(), ...],
+                     num_epochs=10)
+
+# Class API usage. Same as our current TorchTrainer.
+trainer = Trainer(training_operator_cls, num_workers=“auto”, backend=...)
+trainer.init_operator(use_ddp=True, use_fp16=False, use_dist_sampler=...)
```

Comment:
Maybe you can to like:

torch_trainer = Trainer.to_torch_trainer(operator, etc)